src/Microsoft.Extensions.Diagnostics.HealthChecks/DefaultHealthCheckService.cs

```diff
@@ -0,0 +1,149 @@
+            IServiceScopeFactory scopeFactory,
+            IOptions<HealthCheckServiceOptions> options,
+            ILogger<DefaultHealthCheckService> logger)
+        {
+            _scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            // We're specifically going out of our way to do this at startup time. We want to make sure you
+            // get any kind of health-check related error as early as possible. Waiting until someone
+            // actually tries to **run** health checks would be real baaaaad.
+             ValidateRegistrations(_options.Value.Registrations);
+        }
+        public override async Task<HealthReport> CheckHealthAsync(
+            Func<HealthCheckRegistration, bool> predicate,
+            CancellationToken cancellationToken = default)
+        {
+            var registrations = _options.Value.Registrations;
+
+            using (var scope = _scopeFactory.CreateScope())
+            {
+                var context = new HealthCheckContext();
+                var entries = new Dictionary<string, HealthReportEntry>(StringComparer.OrdinalIgnoreCase);
+
+                foreach (var registration in registrations)
+                {
+                    if (predicate != null && !predicate(registration))
+                    {
+                        continue;
+                    }
+
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var healthCheck = registration.Factory(scope.ServiceProvider);
+
+                    // If the health check does things like make Database queries using EF or backend HTTP calls,
+                    // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
+                    using (_logger.BeginScope(new HealthCheckLogScope(registration.Name)))
+                    {
+                        var stopwatch = ValueStopwatch.StartNew();
+                        context.Registration = registration;
+
+                        Log.HealthCheckBegin(_logger, registration);
+                        
+                        HealthReportEntry entry;
+                        try
+                        {
+                            var result = await healthCheck.CheckHealthAsync(context, cancellationToken);
+
+                            entry = new HealthReportEntry(
+                                result.Result ? HealthStatus.Healthy : registration.FailureStatus,
+                                result.Description,
+                                result.Exception,
+                                result.Data);
+
+                            Log.HealthCheckEnd(_logger, registration, entry, stopwatch.GetElapsedTime());
+                        }
+                        catch (Exception ex)
```

Comment:
Do you want to handle OperationCancelledException differently?

Reply:
Yeah, we have an issue for that https://github.com/aspnet/Diagnostics/issues/480